src/main/kotlin/org/jitsi/nlj/rtp/codec/vp9/Vp9Packet.kt

```diff
@@ -0,0 +1,285 @@
+
+    /** End of VP9 picture is the marker bit. Note frame/picture distinction. */
+    /* TODO: not sure this should be the override from [ParsedVideoPacket] */
+    val isEndOfPicture: Boolean
+        /** This uses [get] rather than initialization because [isMarked] is a var. */
+        get() = isMarked
+
+    val hasLayerIndices =
+        DePacketizer.VP9PayloadDescriptor.hasLayerIndices(buffer, payloadOffset, payloadLength)
+
+    val hasPictureId =
+        DePacketizer.VP9PayloadDescriptor.hasPictureId(buffer, payloadOffset, payloadLength)
+
+    val hasExtendedPictureId =
+        DePacketizer.VP9PayloadDescriptor.hasExtendedPictureId(buffer, payloadOffset, payloadLength)
+
+    val hasTL0PICIDX =
+        DePacketizer.VP9PayloadDescriptor.hasTL0PICIDX(buffer, payloadOffset, payloadLength)
+
+    val hasScalabilityStructure =
+        DePacketizer.VP9PayloadDescriptor.hasScalabilityStructure(buffer, payloadOffset, payloadLength)
+
+    val isUpperLevelReference =
+        DePacketizer.VP9PayloadDescriptor.isUpperLevelReference(buffer, payloadOffset, payloadLength)
+
+    val isInterPicturePredicted =
+        DePacketizer.VP9PayloadDescriptor.isInterPicturePredicted(buffer, payloadOffset, payloadLength)
+
+    private var _TL0PICIDX =
+        TL0PICIDX ?: DePacketizer.VP9PayloadDescriptor.getTL0PICIDX(buffer, payloadOffset, payloadLength)
+
+    var TL0PICIDX: Int
+        get() = _TL0PICIDX
+        set(newValue) {
+            _TL0PICIDX = newValue
+            if (newValue != -1 && !DePacketizer.VP9PayloadDescriptor.setTL0PICIDX(
+                    buffer, payloadOffset, payloadLength, newValue)) {
+                logger.cwarn { "Failed to set the TL0PICIDX of a VP9 packet." }
+            }
+        }
+
+    private var _pictureId =
+        pictureId ?: DePacketizer.VP9PayloadDescriptor.getPictureId(buffer, payloadOffset, payloadLength)
+
+    var pictureId: Int
+        get() = _pictureId
+        set(newValue) {
+            _pictureId = newValue
+            if (!DePacketizer.VP9PayloadDescriptor.setExtendedPictureId(
+                    buffer, payloadOffset, payloadLength, newValue)) {
+                logger.cwarn { "Failed to set the picture id of a VP9 packet." }
+            }
+        }
+
+    /* TODO: avoid recomputing these on clone */
+
+    val temporalLayerIndex: Int =
+        DePacketizer.VP9PayloadDescriptor.getTemporalLayerIndex(buffer, payloadOffset, payloadLength)
+
+    val spatialLayerIndex: Int =
+        DePacketizer.VP9PayloadDescriptor.getSpatialLayerIndex(buffer, payloadOffset, payloadLength)
+
+    val isSwitchingUpPoint: Boolean =
+        DePacketizer.VP9PayloadDescriptor.isSwitchingUpPoint(buffer, payloadOffset, payloadLength)
+
+    val usesInterLayerDependency: Boolean =
+        DePacketizer.VP9PayloadDescriptor.usesInterLayerDependency(buffer, payloadOffset, payloadLength)
+
+    fun getScalabilityStructure(
+        eid: Int = 0,
+        baseFrameRate: Double = 30.0
```

Comment:
The caller doesn't seem to override the `baseFrameRate`. Can we snif this on the wire or is it something that we get from signaling?

Reply:
We can measure the actual frame rate as received, but it's not signaled anywhere (either in signaling or in-band).  (The actual frame rate can vary a lot because of things like camera auto-exposure.)

For our current system and current browsers I think assuming 30 fps as the base rate is probably fine.  If someone starts sending 60 fps video it may need to be revisited.